numba/cuda/tests/cudapy/test_vectorize.py

```diff
@@ -227,6 +228,36 @@ def bar(x):
 
         self.assertEqual(bar.__name__, 'bar')
 
+    def test_no_transfer_for_device_data(self):
+        noise = np.random.randn(1, 3, 64, 64).astype(np.float32)
+        noise = cuda.to_device(noise)
+
+        # A mock of a CUDA function that always raises a CudaAPIError
+        def raising_transfer(*args, **kwargs):
+            raise CudaAPIError(999, 'Transfer not allowed')
+
+        old_HtoD = getattr(driver, 'cuMemcpyHtoD', None)
+        old_DtoH = getattr(driver, 'cuMemcpyDtoH', None)
+
+        setattr(driver, 'cuMemcpyHtoD', raising_transfer)
+        setattr(driver, 'cuMemcpyDtoH', raising_transfer)
```

Comment:
Suggestion for future (when not in a hurry), try `unittest.mock`

Reply:
I do use it in some other places but I wasn't quite sure I could reliably replace the function on the driver singleton and remove it again afterwards if it wasn't present before with `mock`.